087-external-certificate-manager.md

```diff
@@ -0,0 +1,241 @@
+This proposal will specifically describe how this would work for cert-manager, however the user API for configuration will be written in a way that does not prevent other external certificate managers being added in the future.
+
+The proposal makes a few assumptions:
+* Strimzi will not be responsible for installing cert-manager, but we will document the supported versions of cert-manager that we have tested with.
+* Strimzi will not be responsible for creating `Issuer` or `ClusterIssuer` custom resources.
+* Strimzi will not be responsible for requesting a CA certificate be issued, it will only interact with cert-manager to request end-entity certificates are issued.
+* Strimzi will create `Certificate` custom resources (to request a new end-entity certificate is issued) and will allow the user to influence the contents of these resources by exposing options in the `Kafka` custom resource.
+* Strimzi will not directly interact with the lower level `CertificateRequest` and `CertificateSigningRequests` custom resources.
+* When Strimzi creates a `Certificate` custom resource, cert-manager will issue the certificate within a reasonable amount of time such that Strimzi can wait during the reconciliation.
+* Users will provide to the Strimzi Cluster Operator the CA certificates it must trust for the current issuer via a Kubernetes Secret.
+
+### API
+
+The existing `spec.clusterCa` and `spec.clientsCa` fields will be extended to add a new property `type`:
+
+```yaml
+spec:
+  clusterCa:
+    validityDays: <integer> # notBefore=now, notAfter=now + validityDays
+    generateCertificateAuthority: <boolean>
+    generateSecretOwnerReference: <boolean>
+    renewalDays: <integer> # days before notAfter when we should start renewal
+    certificateExpirationPolicy: <renew-certificate|replace-key>
+    type: <internal|cert-manager.io> # (1)
+    certManagerIssuerRef: # (2)
+      name: <string>
+      kind: <Issuer|ClusterIssuer>
+      group: <string> # cert-manager.io by default
+    publicCert: # (3)
+      secretName: <string>
+      certificate: <string>
+```
+
+1. The`type` property will default to `internal` when not set and will use the existing behaviour, allowing backwards compatibility.
+   The option `cert-manager.io` will only be valid if `generateCertificateAuthority` is set to `false`.
+2. The property `certManagerIssuerRef` will only be used by Strimzi if `type` is set to `cert-manager.io`.
+  The `name`, `kind`, and `group` properties will be copied over into the `Certificate` custom resource Strimzi creates.
+3. The property `publicCert` will only be used by Strimzi if `type` is set to `cert-manager.io`.
+  The `secretName` and `certificate` properties will be used to locate the CA public certificate that must be trusted by Strimzi components in order to trust the end-entity certificates that cert-manager issues.
```

Comment:
I guess the way this will be implemented is that we will have different classes with different options (as we do for example for authorization, authentication ... and most of the other _typed_ APIs). So each type will have its own fields.

So maybe instead of a long list of fields that can be used one or the other type, you can simply provide two YAMLs for the two different types we will support today?

Alternative is to do what we have for listeners where we have a single class and the validation needs to be done in the operator code. But the main reason for that is that the listeners are defined in an array which is not the case for the CAs. So not sure there would be any benefit of doing that.

Reply:
@scholzj do you mean have two separate yamls in the proposal to make it clearer? Or restructuring the yaml to put the fields for the specific type under that type somehow?

Reply:
I guess if you wanna keep all the options as you suggest in https://github.com/strimzi/proposals/pull/135#discussion_r1926764306 the example is probably fine and you can ignore this.